lab2/part1/myrouter_part1.py

```diff
@@ -0,0 +1,149 @@
+'''
+Title:          myrouter_part1
+Description:    Project 2, part 1 - implement basic ARP handling in a simulated router
+Authors:        David Billmire, Shang Ding
+Course:         UWCS 640, Section 2, Fall 2019
+'''
+
+import sys
+import os
+import time
+
+from switchyard.lib.packet.util import *
+from switchyard.lib.userlib import *
+
+'''
+Class:          ARPContext
+Description:    Implements basic ARP functions, including IP to MAC translation and table management
+                ARP Context is iterable and supports direct get/set of IPv4 -> MAC mapping
+'''
+class ARPContext:
+    def __init__(self):
+        self.map = {}
+        return
+    
+    def __iter__(self):
+        return iter(self.map)
+
+    def __getitem__(self, item):
+        if item in self.map and isinstance(self.map[item], ARPContext.ARPEntry):
+            return self.map[item].mac_addr
+        else:
+            return None
+
+    def __setitem__(self, ip_addr, mac_addr):
+        self.add_mapping(ip_addr, mac_addr)
+        return
+
+    #Add or updates an IP to MAC address mapping
+    #  ip_addr      IPv4Address
+    #  mac_addr     EthAddr
+    def add_mapping(self, ip_addr, mac_addr):
+        if isinstance(ip_addr, IPv4Address) and isinstance(mac_addr, EthAddr):
+            self.map[ip_addr] = ARPContext.ARPEntry(mac_addr)
+        return
+
+    #Process an ARP requests in the provided packet. Return a reply packet
+    #if the target IP address is in our table
+    #   pkt         Arp packet to process
+    def arp_request(self, pkt: Arp):
+        reply_ip  = pkt.targetprotoaddr
+        reply_mac = self.map[reply_ip].mac_addr
```

Comment:
The `self.map[reply_ip].mac_addr` line can crash the program with a `KeyError` if `reply_ip` is not in `self.map`.

I don't understand how/why we pass the class-provided `stage1_tests.srpy` tests which asserts the case of `ARP request for 10.10.1.2 should arrive on router-eth1, but the router should not respond.` 

We should catch the case of a missing key (e.g. the IP isn't in the map) and correspondingly not reply based on my understanding of the instruction:

>For each ARP request, you should determine whether the targetprotoaddr field (IP address destination) in the ARP header is an IP address assigned to one of the interfaces on your router.

>If the destination IP address is one that is assigned to an interface of your Router, you should create and send an appropriate ARP reply. (If the destination IP address is not assigned to one of the router's interfaces, you should not respond with an ARP reply, even if you have enough information to do so. Just drop this packet)